File: Hnefatafl/HnefataflBoard.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <cstdlib>
#include "HnefataflBoard.h"

HnefataflBoard::HnefataflBoard() {
    board.resize(11, std::vector<Piece>(11, EMPTY));
    resetBoard();
}

void HnefataflBoard::resetBoard() {
    std::vector<std::vector<Piece>> initialPosition = {
        {EMPTY,    EMPTY,    EMPTY,     ATTACKER,  ATTACKER,  ATTACKER,  ATTACKER,  ATTACKER,  EMPTY,   EMPTY,     EMPTY},
        {EMPTY,    EMPTY,    EMPTY,     EMPTY,     EMPTY,     ATTACKER,  EMPTY,     EMPTY,     EMPTY,   EMPTY,     EMPTY},
        {EMPTY,    EMPTY,    EMPTY,     EMPTY,     EMPTY,     EMPTY,     EMPTY,     EMPTY,     EMPTY,   EMPTY,     EMPTY},
        {ATTACKER, EMPTY,    EMPTY,     EMPTY,     EMPTY,     DEFENDER,  EMPTY,     EMPTY,     EMPTY,   EMPTY,     ATTACKER},
        {ATTACKER, EMPTY,    EMPTY,     EMPTY,     DEFENDER,  DEFENDER,  DEFENDER,  EMPTY,     EMPTY,   EMPTY,     ATTACKER},
        {ATTACKER, ATTACKER, EMPTY,     DEFENDER,  DEFENDER,  KING,      DEFENDER,  DEFENDER,  EMPTY,   ATTACKER,  ATTACKER},
        {ATTACKER, EMPTY,    EMPTY,     EMPTY,     DEFENDER,  DEFENDER,  DEFENDER,  EMPTY,     EMPTY,   EMPTY,     ATTACKER},
        {ATTACKER, EMPTY,    EMPTY,     EMPTY,     EMPTY,     DEFENDER,  EMPTY,     EMPTY,     EMPTY,   EMPTY,     ATTACKER},
        {EMPTY,    EMPTY,    EMPTY,     EMPTY,     EMPTY,     EMPTY,     EMPTY,     EMPTY,     EMPTY,   EMPTY,     EMPTY},
        {EMPTY,    EMPTY,    EMPTY,     EMPTY,     EMPTY,     ATTACKER,  EMPTY,     EMPTY,     EMPTY,   EMPTY,     EMPTY},
        {EMPTY,    EMPTY,    EMPTY,     ATTACKER,  ATTACKER,  ATTACKER,  ATTACKER,  ATTACKER,  EMPTY,   EMPTY,     EMPTY}
    };

    board = initialPosition;
    currentPlayer = PLAYER_ATTACKER;
    turnNum = 1;
    srand(time(0));
}



void HnefataflBoard::printBoard() const {
    int BOARD_SIZE = 11;
    const std::string RESET = "\033[0m";
    const std::string RED = "\033[31m";
    const std::string GREEN = "\033[32m";
    const std::string BOLD_BLUE = "\033[1;34m";

    std::cout << "Turn number: " << HnefataflBoard::turnNum << std::endl;
    // Print the column labels (A-K) with extra spacing
    std::cout << "   ";
    for (char colLabel = 'A'; colLabel <= 'K'; ++colLabel) {
        std::cout << " " << colLabel << " ";
    }
    std::cout << std::endl;
    std::cout << std::endl;

    for (int i = BOARD_SIZE - 1; i >= 0; i--) { 
        // Print the row labels (1-11) with extra spacing
        if (i < 9) { // Single-digit row numbers
            std::cout << (i + 1) << "  ";
        } else {    // Double-digit row numbers
            std::cout << (i + 1) << " ";
        }

        for (int j = 0; j < BOARD_SIZE; j++) {
            switch (board[i][j]) {
                case EMPTY: std::cout << " . "; break;
                case ATTACKER: std::cout << RED << " A " << RESET; break;
                case DEFENDER: std::cout << GREEN << " D " << RESET; break;
                case KING: std::cout << BOLD_BLUE << " K " << RESET; break;
            }
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
    std::cout << "   ";
    for (char colLabel = 'A'; colLabel <= 'K'; ++colLabel) {
        std::cout << " " << colLabel << " ";
    }
    std::cout << std::endl;
}

HnefataflBoard::Player HnefataflBoard::getCurrentPlayer() const {
    return currentPlayer;
}


HnefataflBoard::Move::Move(int sX, int sY, int eX, int eY) 
    : startX(sX), startY(sY), endX(eX), endY(eY) { }

std::vector<HnefataflBoard::Move> HnefataflBoard::generateLegalMoves(HnefataflBoard::Player player) const {
    std::vector<HnefataflBoard::Move> legalMoves;

    for (int i = 0; i < 11; i++) {
        for (int j = 0; j < 11; j++) {
            if ((board[i][j] == ATTACKER && player == PLAYER_ATTACKER) ||
                (board[i][j] == DEFENDER && player == PLAYER_DEFENDER) ||
                (board[i][j] == KING && player == PLAYER_DEFENDER)) {

                // Check in each direction: up, down, left, right
                int dx[] = {0, 0, -1, 1};
                int dy[] = {-1, 1, 0, 0};
                for (int dir = 0; dir < 4; dir++) {
                    int x = i + dx[dir], y = j + dy[dir];
                    while (x >= 0 && x < 11 && y >= 0 && y < 11 && board[x][y] == EMPTY) {
                        legalMoves.emplace_back(i, j, x, y);
                        x += dx[dir];
                        y += dy[dir];
                    }
                }
            }
        }
    }

    return legalMoves;
}

bool HnefataflBoard::makeMove(HnefataflBoard::Player player, int startX, int startY, int endX, int endY) {
    // Check if it's the correct player's turn
    if (player != currentPlayer) {
        std::cout << "It's not this player's turn." << std::endl;
        return false;
    }

    // Check if the starting square contains the player's piece
    if ((player == HnefataflBoard::PLAYER_ATTACKER && board[startX][startY] != HnefataflBoard::ATTACKER) ||
        (player == HnefataflBoard::PLAYER_DEFENDER && (board[startX][startY] != HnefataflBoard::DEFENDER && board[startX][startY] != HnefataflBoard::KING))) {
        std::cout << "The starting square does not contain the player's piece." << std::endl;
        return false;
    }

    if (!isMoveLegal(startX, startY, endX, endY)) {
        std::cout << "Move is diagonal, or passes through a piece. Try again." << std::endl;
        return false;
    }

    // If all checks pass, check if the destination is a corner square and if the moving piece isn't the king
    if ((endX == 0 || endX == 10) && (endY == 0 || endY == 10) && board[startX][startY] != KING) {
        std::cout << "Only the king can move to the corner squares." << std::endl;
        return false;
    }
    // Check if the ending square is empty
    if (board[endX][endY] != HnefataflBoard::EMPTY) {
        std::cout << "The destination square is not empty." << std::endl;
        return false;
    }

    // If all checks pass, move the piece and switch the current player
    board[endX][endY] = board[startX][startY];
    board[startX][startY] = HnefataflBoard::EMPTY;

    std::cout << "Piece moved from " << startX << ", " << startY << " to " << endX << ", " << endY << std::endl;

    evaluateCaptures(endX, endY);

    currentPlayer = (currentPlayer == HnefataflBoard::PLAYER_ATTACKER) ? HnefataflBoard::PLAYER_DEFENDER : HnefataflBoard::PLAYER_ATTACKER;

    HnefataflBoard::turnNum += 1;
    return true;
}


HnefataflBoard::Move HnefataflBoard::getRandomMove(HnefataflBoard::Player player) const {
    std::vector<Move> legalMoves = generateLegalMoves(player);

    if (legalMoves.empty()) {
        throw std::runtime_error("No legal moves available for the player!");
    }

    // Pick and return a random move
    int randomIndex = std::rand() % legalMoves.size();
    return legalMoves[randomIndex];
}

bool HnefataflBoard::performRandomMove(HnefataflBoard::Player player) {
    HnefataflBoard::Move randomMove = getRandomMove(player);
    return makeMove(player, randomMove.startX, randomMove.startY, randomMove.endX, randomMove.endY);
}

bool HnefataflBoard::checkForWin() {
    // Check for the king's victory
    for (int i = 0; i < 11; i++) {
        for (int j = 0; j < 11; j++) {
            if (board[i][j] == HnefataflBoard::KING) {
                if (isKingAtCorner(i, j)) {
                    std::cout << "Defenders win! The king has escaped!" << std::endl;
                    return true;
                }
            }
        }
    }

    // Check for the attackers' victory
    if (isKingCaptured()) {
        std::cout << "Attackers win! The king has been captured!" << std::endl;
        return true;
    }

    return false; // No one has won yet
}


bool HnefataflBoard::parseSquare(const std::string& squareName, int& x, int& y) {
    if (squareName.size() < 2 || squareName.size() > 3) return false;

    char colLetter = std::toupper(squareName[0]);
    if (colLetter < 'A' || colLetter > 'K') return false;

    int rowNum;
    try {
        rowNum = std::stoi(squareName.substr(1));
    } catch (std::exception&) {
        return false;
    }

    if (rowNum < 1 || rowNum > 11) return false;

    y = colLetter - 'A';                // Convert column letter to index
    x = rowNum - 1;              // Convert row number to index

    return true;
}


bool HnefataflBoard::isKingAtCorner(int x, int y) const {
    return (x == 0 || x == 10) && (y == 0 || y == 10);
}

bool HnefataflBoard::isKingCaptured() const {
    for (int i = 0; i < 11; i++) {
        for (int j = 0; j < 11; j++) {
            if (board[i][j] == KING) {
                // Check for top edge or ATTACKER piece above
                bool topHostile = (i == 0) || (board[i - 1][j] == ATTACKER);

                // Check for bottom edge or ATTACKER piece below
                bool bottomHostile = (i == 10) || (board[i + 1][j] == ATTACKER);

                // Check for left edge or ATTACKER piece to the left
                bool leftHostile = (j == 0) || (board[i][j - 1] == ATTACKER);

                // Check for right edge or ATTACKER piece to the right
                bool rightHostile = (j == 10) || (board[i][j + 1] == ATTACKER);

                std::cout << topHostile << " " << bottomHostile << " " << leftHostile << " " << rightHostile << std::endl;
                // If all sides around the king are hostile, then the king is captured
                return topHostile && bottomHostile && leftHostile && rightHostile;
            }
        }
    }
    return false; // This should never be reached, but it's here just in case.
}

HnefataflBoard::Piece HnefataflBoard::getPieceAt(int x, int y) const {
    // You might want to add boundary checks to ensure x and y are valid
    if(x < 0 || x >= board.size() || y < 0 || y >= board[0].size()) {
        // Return a default value or throw an exception
        return EMPTY; // or throw std::out_of_range("Invalid board coordinates");
    }
    return board[x][y];
}


HnefataflBoard::Player currentPlayer = HnefataflBoard::PLAYER_ATTACKER;  // We'll assume the attacker goes first.


bool HnefataflBoard::isMoveLegal(int startX, int startY, int endX, int endY) const {
    if (startX != endX && startY != endY) {
        return false; // Diagonal move is illegal
    }
    
    // Horizontal move
    if (startX == endX) {
        int step = (startY < endY) ? 1 : -1;
        for (int y = startY + step; y != endY; y += step) {
            if (board[startX][y] != HnefataflBoard::EMPTY) {
                return false; // Cannot move past another piece
            }
        }
    }

    // Vertical move
    if (startY == endY) {
        int step = (startX < endX) ? 1 : -1;
        for (int x = startX + step; x != endX; x += step) {
            if (board[x][startY] != HnefataflBoard::EMPTY) {
                return false; // Cannot move past another piece
            }
        }
    }

    return true;
}

void HnefataflBoard::evaluateCaptures(int x, int y) {
    Piece movedPiece = board[x][y];
    Piece opponentPiece = (movedPiece == ATTACKER) ? DEFENDER : ATTACKER;

    // Function to check if the position is hostile
    auto isHostile = [this, movedPiece](int x, int y) {
        return ((x == 0 && y == 0) || (x == 10 && y == 0) || (x == 0 && y == 10) || (x == 10 && y == 10)) || board[x][y] == movedPiece;
    };

    // Check horizontal capture
    if (x > 1 && board[x - 1][y] == opponentPiece && isHostile(x - 2, y)) {
        board[x - 1][y] = EMPTY;
        std::cout << "### Piece captured on " << x - 1 << ", " << y << std::endl;
    }
    if (x < 9 && board[x + 1][y] == opponentPiece && isHostile(x + 2, y)) {
        board[x + 1][y] = EMPTY;
        std::cout << "### Piece captured on " << x + 1 << ", " << y << std::endl;
    }

    // Check vertical capture
    if (y > 1 && board[x][y - 1] == opponentPiece && isHostile(x, y - 2)) {
        board[x][y - 1] = EMPTY;
        std::cout << "### Piece captured on " << x << ", " << y - 1 << std::endl;
    }
    if (y < 9 && board[x][y + 1] == opponentPiece && isHostile(x, y + 2)) {
        board[x][y + 1] = EMPTY;
        std::cout << "### Piece captured on " << x << ", " << y + 1 << std::endl;
    }
}

File: Hnefatafl/Python/HnefataflBoard.py
import numpy as np
from random import randint, choice, shuffle
from copy import deepcopy

EMPTY = 0
ATTACKER = 1
DEFENDER = -1
KING = -2

corners_arr = np.array([[0, 0], [10, 0], [0, 10], [10,10]])
corners = [[0, 0], [10, 0], [0, 10], [10,10]]
offsets = np.array([[-1,0], [0,-1], [1,0], [0,1]])

piece_names = {0: "·", 1: "♖", -1: "♜", -2: "♚"}
player_names = {0: "None", 1: "Attacker", -1: "Defender"}


class Board:
    def __init__(self):
        self.board = np.array(
            [[ 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0],
            [ 0 ,0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
            [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [ 1, 0, 0, 0, 0,-1, 0, 0, 0, 0, 1],
            [ 1, 0, 0, 0,-1,-1,-1, 0, 0, 0, 1],
            [ 1, 1, 0,-1,-1,-2,-1,-1, 0, 1, 1],
            [ 1, 0, 0, 0,-1,-1,-1, 0, 0, 0, 1],
            [ 1, 0, 0, 0, 0,-1, 0, 0, 0, 0, 1],
            [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
            [ 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
            [ 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0]],
            dtype=np.int8)
        
        self.current_player = ATTACKER
        self.turn = 1
        self._legal_moves = None
        self.won = 0


    def print_board(self):
        board = self.board
        print(f"Turn {self.turn}, {player_names[self.current_player]} to move.")
        print(f"Current score: {self.get_current_score()}")
        printstring = ""
        printstring += f"   A B C D E F G H I J K\n"
        for x in range(11):
            printstring += f"{x+1:2d} "
            for y in range(11):
                if x in [0, 10] and y in [0, 10] and board [x,y] == EMPTY:
                    printstring += "○ "
                elif x == 5 and y == 5 and board [x,y] == EMPTY:
                    printstring += "□ "
                else:
                    printstring += piece_names[board[x,y]] + " "
            printstring += f" {x+1:d}"
            printstring += "\n"
        printstring += f"   A B C D E F G H I J K"
        print(printstring)


    @property
    def legal_moves(self):
        if self._legal_moves is None:
            player = self.current_player
            board = self.board
            self._legal_moves = []
            for x in range(11):
                for y in range(11):
                    piece = board[x,y]
                    if (player == ATTACKER and board[x,y] == ATTACKER) or (player == DEFENDER and (board[x,y] == DEFENDER or board[x,y] == KING)):
                        for x_new in range(x+1, 11, 1):
                            if board[x_new, y] == EMPTY:
                                if x_new != 5 or y != 5:  # Throne can never be moved to, but can be passed through.
                                    if [x_new, y] not in corners or piece == KING:  # Only king can move to the corners
                                        self._legal_moves.append([[x,y],[x_new,y]])
                            else:
                                break
                        for x_new in range(x-1, -1, -1):
                            if board[x_new, y] == EMPTY:
                                if x_new != 5 or y != 5:
                                    if [x_new, y] not in corners or piece == KING:
                                        self._legal_moves.append([[x,y],[x_new,y]])
                            else:
                                break
                        for y_new in range(y+1, 11, 1):
                            if board[x, y_new] == EMPTY:
                                if x != 5 or y_new != 5:
                                    if [x, y_new] not in corners or piece == KING:
                                        self._legal_moves.append([[x,y],[x,y_new]])
                            else:
                                break
                        for y_new in range(y-1, -1, -1):
                            if board[x, y_new] == EMPTY:
                                if x != 5 or y_new != 5:
                                    if [x, y_new] not in corners or piece == KING:
                                        self._legal_moves.append([[x,y],[x,y_new]])
                            else:
                                break

        return self._legal_moves


    def make_move(self, move, verbose=True):
        player = self.current_player
        board = self.board
        move_from, move_to = move
        if (player == ATTACKER and board[*move_from] != 1) or (player == DEFENDER and board[*move_from] not in [DEFENDER, KING]):
            raise ValueError(f"Player {player} tried moving piece {piece_names[board[*move_from]]}")
        if move not in self.legal_moves:
            raise ValueError(f"Move {move} is not a legal move.")
        board[*move_to] = board[*move_from]
        board[*move_from] = EMPTY

        for offset in offsets:  # Consider all neighboring squares for potential captures.
            if 0 <= move_to[0]+offset[0] <= 10 and 0 <= move_to[1]+offset[1] <= 10:  # If neighboring square is inside board.
                if board[move_to[0]+offset[0], move_to[1]+offset[1]]*player < 0 and board[move_to[0]+offset[0], move_to[1]+offset[1]] != KING:  # If neighboring piece is enemy but not king.
                    # For a enemy neighboring piece, look at the square one further in the same direction.
                    double_neighbor_x = move_to[0]+offset[0]*2
                    double_neighbor_y = move_to[1]+offset[1]*2
                    # Check if that square is "hostile", aka either: inside the board and has allied piece, or a corner square. 
                    if (0 <= double_neighbor_x <= 10 and 0 <= double_neighbor_y <= 10 and board[double_neighbor_x,double_neighbor_y]*player > 0)\
                        or (double_neighbor_x in [0, 10] and double_neighbor_y in [0,10]):
                        capture_x, capture_y = move_to[0]+offset[0], move_to[1]+offset[1]
                        if verbose:
                            print(f"##### Piece captured on {capture_x}, {capture_y}. #####")
                        board[capture_x,capture_y] = EMPTY
        if verbose:
            print(f"Making move {move[0]} to {move[1]}.")

        self.current_player *= -1  # Switching current player.
        self._legal_moves = None  # Resetting legal moves, as they don't apply to new board state.


    def evaluate_win(self, verbose=True):
        board = self.board
        king_pos = np.argwhere(board == KING)[0]
        if (king_pos == corners_arr[0]).all() or (king_pos == corners_arr[1]).all() or (king_pos == corners_arr[2]).all() or (king_pos == corners_arr[3]).all():
            if verbose:
                print("King in corner.")
            self.won = DEFENDER
            return DEFENDER
        king_surrounded = True
        for offset in offsets:
            pos = king_pos+offset
            if not ((0 < pos)*(pos < 11)).all():
                king_surrounded = False
            else:
                if board[*pos] != ATTACKER:
                    king_surrounded = False
        if king_surrounded:
            if verbose:
                print("King surrounded.")
            self.won = ATTACKER
            return ATTACKER
        if np.sum(board == DEFENDER) == 0:
            if verbose:
                print("No pieces left for defender.")
            self.won = ATTACKER
            return ATTACKER
        return 0


    def get_current_score(self):
        board = self.board
        if self.won != 0:
            return np.inf*self.won
        return np.sum(board == 1) - np.sum(board==-1)*2


    def play_random_move(self):
        # Plays a random move for the current player.
        legal_moves = self.legal_moves
        chosen_move = choice(legal_moves)
        self.make_move(chosen_move)

    def play_AI_1(self):
        legal_moves = self.legal_moves.copy()
        shuffle(legal_moves)  # Shuffle it so it plays a random best move, not just always the same best move.
        scores = np.zeros((len(legal_moves)))
        for i in range(len(legal_moves)):
            new_board = deepcopy(self)
            new_board.make_move(legal_moves[i], verbose=False)
            new_board.evaluate_win(verbose=False)
            scores[i] = new_board.get_current_score()
        best_move_idx = np.argmax(scores*self.current_player)
        self.make_move(legal_moves[best_move_idx])